### ISSUE TITLE
Net 4.0 with Dapper and Ms Sql 2000 

### ISSUE THREAD

**Issue by postondemand** [StackExchange/Dapper#434]
As we know Ms Sql 2000 does not support MultipleActiveResultSets.

Can i use Dapper with async Task without exceptions :
"There is already an open DataReader associated with this Command which must be closed first."

Thank you.

**Comment by mgravell** [comment]
Absolutely! You just need to ensure one operation completes before you try
to do another one. It is not uncommon for MARS to not be enabled in the
connection string, so it is entirely expected that it needs to work without
MARS. In simple cases, this means waiting or awaiting for Query (etc)
methods to complete. If you have used "buffered: false"​, then you will
also need to make sure that you enumerate (foreach) over any data *before*
issuing any additional commands.

**Comment by postondemand** [comment]
i will I look at this "buffered: false"

my code example

  ```csharp
private async void Form_Load(object sender, EventArgs e)
{
	var sql1 = "select from Tab1";
	var sql2 = "select from Tab2";
	var sql3 = "select from Tab3";

	await Task.Factory.StartNew(() => FillComboBoxWithData(this.cbo1, sql1));
	await Task.Factory.StartNew(() => FillComboBoxWithData(this.cbo2, sql2));
	await Task.Factory.StartNew(() => FillComboBoxWithData(this.cbo3, sql3));
}

public static async Task FillComboBoxWithData(ComboBox comboBox, string sql)
{
	try
	{
		var data = await Task.Factory.StartNew(() => SqlConn.Query<IdName>(sql));
		var d1 = data.ToNonNullList();

		comboBox.DataSource = d1;

		comboBox.DisplayMember = "Name";
		comboBox.ValueMember = "Id";
		comboBox.SelectedItem = null;
	}
	catch (Exception ex)
	{
		Console.Write(ex.ToString());
	}
}

**Comment by mgravell** [comment]
`buffered: true` is the default, and avoids a lot of problems to do with the scenario you describe; you shouldn't need to do anything there - it was actually more of a warning: "if you're using `buffered: false`, you need to be more careful". You're not using that in the code you show, so: don't try and add it now!

As for your code... it looks like it should work, although I'm very confused as to why you're using `Task.Factory.StartNew` everywhere; you shouldn't have to do that, and indeed: it is not ideal to do so.

I also can't see how / where `SqlConn` is defined, so for all I know it is indeed being accessed concurrently. However, to rewrite your code more idiomatically:

    private async void Form_Load(object sender, EventArgs e)
    {
        var sql1 = "select from Tab1";
        var sql2 = "select from Tab2";
        var sql3 = "select from Tab3";

        await FillComboBoxWithData(this.cbo1, sql1);
        await FillComboBoxWithData(this.cbo2, sql2);
        await FillComboBoxWithData(this.cbo3, sql3);
    }

    public static async Task FillComboBoxWithData(ComboBox comboBox, string sql)
    {
        try
        {
            var data = await SqlConn.QueryAsync<IdName>(sql).AsList();

            comboBox.DataSource = data;
            comboBox.DisplayMember = "Name";
            comboBox.ValueMember = "Id";
            comboBox.SelectedItem = null;
        }
        catch (Exception ex)
        {
            Console.Write(ex.ToString());
        }
    }

**Comment by postondemand** [comment]
@mgravell you are absolutely right.

Task.Factory.StartNew It is completely unnecessary because async / await.

Thank you for your help. I will test that solution tomorrow.

Thank you also for great solution which is dapper. 

Can't code without it :)

**Comment by postondemand** [comment]
i have nuget package

"packages\Dapper.1.42\lib\net40\Dapper.dll"

and QueryAsync is not available

Should i download source code and recompile with some compilation symbol or something?

**Comment by mgravell** [comment]
Hmmm... yeah, dammit, `DbCommand.ExecuteReaderAsync` isn't added until .NET 4.5. Frankly, if you are serious about doing anything `async` that involves database, probably the best advice I can give you is: update to vLatest. 4.0 is 5 years old now - that's a long time in software.

**Comment by mgravell** [comment]
Note: using "async over sync" (the thing involving task factory) is usually a *terrible* idea, and should be avoided. As should "sync over async".

**Comment by postondemand** [comment]
we have old customer, who has a database server from the beginning of sql 2000, and I probably will not live the moment in which they will migrate.
it is also barrier ms server 2003 which supports max only NF 4.0

So I know I need to change my job;)

Thank you for your help.

**Comment by mgravell** [comment]
In that case: do it the old fashioned way. Use a background worker to fetch
the data (Query, etc), then send the data to the UI thread to update the
controls. Using async in a place where async doesn't really exist (4.0
DbCommand) is just a recipe for frustration.

**Comment by postondemand** [comment]
```c#
public class AsyncTest
    {
        public async Task<List<IdNazwa>> GetAsync(string query)
        {
            using (var conn = new SqlConnection(DataBase.ConnString("sa",@"fbd43@!city")))
            {
                using (var cmd = new SqlCommand())
                {
                    cmd.Connection = conn;
                    cmd.CommandText = query;
                    cmd.CommandType = CommandType.Text;

                    await conn.OpenAsync();

                    using (var reader = await cmd.ExecuteReaderAsync())
                    {
                        return reader.Select(this.Builder).ToList();
                    }
                }
            }
        }

        //private helpers
        private IdNazwa Builder(SqlDataReader reader)
        {
            return new IdNazwa
                       {
                           Id = int.Parse(reader["Id"].ToString()),
                           Nazwa = reader["Nazwa"] is DBNull ? null : reader["Nazwa"].ToString()
                       };
        }
    }
-------------------------
Extension method
------------------------

        public static IEnumerable<T> Select<T>(this SqlDataReader reader, Func<SqlDataReader, T> projection)
        {
            while (reader.Read())
            {
                yield return projection(reader);
            }
        }

**Comment by NickCraver** [comment]
Note that we'll likely remove .NET 4.0 support in v2, so I'm inclined to close this one out.

While I understand you're trying to make async work in .NET 4.0, the harsh realities are that it's a) unsupported, and b) quite an adventure...as seen in the history above. IMO, we should close this out since the only answer we can *reasonably* support here is: upgrade to .NET 4.5+

**Comment by NickCraver** [comment]
Since the population (and Microsoft support) of .NET 4.0 has tanked, we have removed it from our libraries across the board. With the VS 2017 tooling, we can't even properly build for it in a multi-target setup. Given the sync/async split has additional cost to maintain, it simply wasn't worth it. Old versions of the library will forever remain available on nuget.org for any consumers stuck on that framework version.

Closing this out to wrap up :)
